### PR TITLE
split out DictProxy for re-use from doveauth

### PR DESCRIPTION
**Pull request by hpk42**
Refactor common dictproxy-base code of metadata and doveauth services
so that it's easier to implement a new dictproxy for #352

### CODE REVIEW

File: chatmaild/src/chatmaild/dictproxy.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return "N\n"
+
+    def handle_iterate(self, parts):
+        # Empty line means ITER_FINISHED.
+        # If we don't return empty line Dovecot will timeout.
+        return "\n"
+
+    def handle_begin_transaction(self, transaction_id, parts):
+        addr = parts[1]
+        self.transactions[transaction_id] = dict(addr=addr, res="O\n")
```
Comment (missytake) [inline chatmaild/src/chatmaild/dictproxy.py:61]: I didn't know trans actions were part of this project? 🙃

File: chatmaild/src/chatmaild/tests/test_doveauth.py [hunk 2]
```diff
@@ -120,28 +121,31 @@ def test_handle_dovecot_request(db, example_config):
...
+    dictproxy = AuthDictProxy(db=db, config=example_config)
     lookup_passdb(db, example_config, "asdf00000@chat.example.org", "q9mr3faue")
     lookup_passdb(db, example_config, "asdf11111@chat.example.org", "q9mr3faue")
     rfile = io.BytesIO(b"H3\t2\t0\t\tauth\nI0\t0\tshared/userdb/")
     wfile = io.BytesIO()
-    handle_dovecot_protocol(rfile, wfile, db, example_config)
+    dictproxy.loop_forever(rfile, wfile)
     lines = wfile.getvalue().decode("ascii").split("\n")
     assert lines[0] == "Oshared/userdb/asdf00000@chat.example.org\t"
     assert lines[1] == "Oshared/userdb/asdf11111@chat.example.org\t"
```
Comment (missytake) [inline chatmaild/src/chatmaild/tests/test_doveauth.py:151]: This test doesn't seem to use `gencreds`?

File: chatmaild/src/chatmaild/tests/test_doveauth.py [hunk 3]
```diff
@@ -120,28 +121,31 @@ def test_handle_dovecot_request(db, example_config):
...
 
 def test_handle_dovecot_protocol(db, example_config):
+    dictproxy = AuthDictProxy(db=db, config=example_config)
     rfile = io.BytesIO(
         b"H3\t2\t0\t\tauth\nLshared/userdb/foobar@chat.example.org\tfoobar@chat.example.org\n"
     )
     wfile = io.BytesIO()
-    handle_dovecot_protocol(rfile, wfile, db, example_config)
+    dictproxy.loop_forever(rfile, wfile)
     assert wfile.getvalue() == b"N\n"
```
Comment (missytake) [inline chatmaild/src/chatmaild/tests/test_doveauth.py:139]: So the `protocol` test tests failing authentication and the `request` test tests for working authentication? Maybe we could rename them so it is more obvious on first glance what they do.

File: chatmaild/src/chatmaild/dictproxy.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    ThreadingMixIn,
+    UnixStreamServer,
+)
+
+
+class DictProxy:
+    def __init__(self):
+        self.transactions = {}
+
+    def loop_forever(self, rfile, wfile):
```
Comment (missytake) [inline chatmaild/src/chatmaild/dictproxy.py:14]: A bit wild, seeing `loop_forever` show up in all the tests. But it is obvious enough what it means I'd say :)

**Review by missytake — APPROVED** [review]

File: chatmaild/src/chatmaild/dictproxy.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            pass
+
+        with ThreadedUnixStreamServer(socket, Handler) as server:
+            try:
+                server.serve_forever()
+            except KeyboardInterrupt:
+                pass
+
+
+class ThreadedUnixStreamServer(ThreadingMixIn, UnixStreamServer):
```
[the remark this answers is not in the document]
Comment (hpk42) [reply chatmaild/src/chatmaild/dictproxy.py]: thanks, replaced.

File: chatmaild/src/chatmaild/tests/test_doveauth.py (hunk 3 above)
Comment (hpk42) [reply chatmaild/src/chatmaild/tests/test_doveauth.py]: renamed test

File: chatmaild/src/chatmaild/tests/test_doveauth.py (hunk 2 above)
Comment (hpk42) [reply chatmaild/src/chatmaild/tests/test_doveauth.py:151]: thanks for noticing, removed.

**Pull request merged**